subflow.go

```diff
@@ -0,0 +1,240 @@
+package multipath
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/getlantern/ema"
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+type pendingAck struct {
+	sz     int
+	sentAt time.Time
+}
+type subflow struct {
+	to   string
+	conn net.Conn
+	mpc  *mpConn
+
+	chClose       chan struct{}
+	closeOnce     sync.Once
+	sendQueue     chan *sendFrame
+	pendingAcks   map[uint64]pendingAck
+	muPendingAcks sync.Mutex
+	probeStart    atomic.Value // time.Time
+	emaRTT        *ema.EMA
+	tracker       StatsTracker
+}
+
+func startSubflow(to string, c net.Conn, mpc *mpConn, clientSide bool, probeStart time.Time, tracker StatsTracker) *subflow {
+	sf := &subflow{
+		to:          to,
+		conn:        c,
+		mpc:         mpc,
+		chClose:     make(chan struct{}),
+		sendQueue:   make(chan *sendFrame),
+		pendingAcks: make(map[uint64]pendingAck),
+		emaRTT:      ema.NewDuration(longRTT, rttAlpha),
+		tracker:     tracker,
+	}
+	go sf.sendLoop()
+	if clientSide {
+		initialRTT := time.Since(probeStart)
+		tracker.UpdateRTT(initialRTT)
+		sf.emaRTT.SetDuration(initialRTT)
+		// pong immediately so the server can calculate the RTT between when it
+		// sends the leading bytes and receives the pong frame.
+		sf.ack(frameTypePong)
+	} else {
+		sf.probeStart.Store(probeStart)
+	}
+	go func() {
+		if err := sf.readLoop(); err != nil {
+			log.Debugf("read loop to %s ended: %v", sf.to, err)
+		}
+	}()
+	return sf
+}
+
+func (sf *subflow) readLoop() (err error) {
+	ch := make(chan *frame)
+	r := byteReader{Reader: sf.conn}
+	go func() {
+		defer close(ch)
+		for {
+			sz, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			fn, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			if sz == 0 {
+				sf.gotACK(fn)
+				continue
+			}
+			log.Tracef("got frame %d from %s with %d bytes", fn, sf.to, sz)
+			buf := pool.Get(int(sz))
+			_, err = io.ReadFull(r, buf)
+			if err != nil {
+				pool.Put(buf)
+				sf.close()
+				return
+			}
+			sf.ack(fn)
+			ch <- &frame{fn: fn, bytes: buf}
+			sf.tracker.OnRecv(sz)
+			select {
+			case <-sf.chClose:
+				return
+			default:
+				// continue
+			}
+		}
+	}()
+	probeTimer := time.NewTimer(randomize(probeInterval))
+	for {
+		select {
+		case frame := <-ch:
+			if frame == nil {
+				return
+			}
+			sf.mpc.recvQueue.add(frame)
+			if !probeTimer.Stop() {
+				<-probeTimer.C
```

Comment:
This is accumulating goroutines though I don't know what goes wrong...

```
1068 @ 0x4038ee0 0x4008017 0x4007cdb 0x452e62e 0x45308ff 0x4069201
#	0x452e62d	github.com/getlantern/multipath.(*subflow).readLoop+0x23d	/Users/joesis/go/pkg/mod/github.com/getlantern/multipath@v0.0.0-20200914055716-a0b4261a18ca/subflow.go:113
#	0x45308fe	github.com/getlantern/multipath.startSubflow.func1+0x2e		/Users/joesis/go/pkg/mod/github.com/getlantern/multipath@v0.0.0-20200914055716-a0b4261a18ca/subflow.go:58
```

Reply:
Sorry I should have updated. It's fixed by resetting the timer after probing (a few lines below) as well.